core: add `Subscriber::try_close`

hawkw opened the pull request:
## Motivation

As discussed in #136, a proposed `Layer` trait for composing
subscribers required should probably _not_ be responsible for
managing ref-counts and determining span closures. Instead, the root
subscriber should be responsible for this, and `Layer`s should be able
to opt into a callback that's notified _when_ a span is closed. 

Adding a callback that's called on closure to `Layer` requires modifying
the `Subscriber` trait to allow indicating when a span closes.

## Solution

This branch attempts to do this without a breaking change, by adding a new
`try_close` trait method to `Subscriber`. `try_close` is identical to
`drop_span` except that it returns a boolean value, set to `true` if the
span has closed. 

The `try_close` default implementation simply returns `false`, so that
if subscribers don't care about tracking ref counts, close notifications
will never be triggered. The default implementation of `drop_span` is
changed to call `self.try_close(...)` and ignore the return value. 

The `drop_span` documentation now indicates that it is "soft deprecated"
similarly to `std::error::Error`'s `description` method. This encourages
subscribers to implement the new API, but isn't a breaking change.
Existing _callers_ of `drop_span` are still able to call it without
causing issues, as the `Layered::drop_span` will always call `try_close`
on the inner subscriber, in order to forward the closure notification to
the layer if the span closes.

jonhoo commented on tracing-core/src/dispatcher.rs:
```diff
@@ -315,13 +315,35 @@ impl Dispatch {
+    /// Notifies the subscriber that a [span ID] has been dropped, and returns
+    /// `true` if there are now 0 IDs referring to that span.
+    ///
+    /// This function is guaranteed to only be called with span IDs that were
```
Should this say "must", not "is guaranteed", since method is called by things that have a `Dispatch`?

jonhoo replied:
I think this comment still holds?

hawkw replied:
ah, whoops --- i forgot this one

jonhoo commented on tracing-core/src/subscriber.rs:
```diff
@@ -283,12 +283,26 @@ pub trait Subscriber: 'static {
+    /// ignores the return value.
+    ///
+    /// [`try_close`]: trait.Subscriber.html#method.try_close
+    fn drop_span(&self, id: span::Id) {
```
Maybe I'm missing something here.. If the implementation is updated to call `try_close` instead of `drop_span`, then existing `Subscriber` implementations that only implement `drop_span` would never see that function be called, since `try_close` does not forward to `drop_span`. That seems like a breaking change, again assuming that we'll now start calling `try_close` instead of `drop_span` in `tracing`.

hawkw replied:
Ah, yeah, that might be an issue. I think `try_close` should default to calling `drop_span` instead?

jonhoo replied:
Possibly, but that is an issue too, no -- what do you return from it?

hawkw replied:
@jonhoo it should return `false`. If the subscriber doesn't have a way of indicating that the span should be closed, then close notifications will not occur. In a world where `try_close` defaults to calling `drop_span`, we should probably deprecate manual impls of drop_span for this reason.

hawkw commented:
@jonhoo I've changed `try_close` to call `drop_span`, as you suggested in https://github.com/tokio-rs/tracing/pull/153#discussion_r301312050, and updated the documentation a bit. It would be great to get your approval on this branch, as I'd like to move forward with this.

jonhoo commented:
@hawkw happy to approve subject to the one comment above.
